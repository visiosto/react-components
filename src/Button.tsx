// Copyright (c) 2021 Visiosto oy
// Licensed under the MIT License

import React from 'react';
import styled from 'styled-components';

import buttonDefaultStyle from './style/buttonDefaultStyle';
import buttonDefaultThinStyle from './style/buttonDefaultThinStyle';
import buttonStyle from './style/buttonStyle';
import buttonThinStyle from './style/buttonThinStyle';

const Span = styled.span`
  ${buttonStyle}
`;

const SpanThin = styled.span`
  ${buttonThinStyle}
`;

const SpanDefault = styled(Span)`
  ${buttonDefaultStyle}
`;

const SpanAccept = styled(Span)`
  background-color: var(--color-accept);
  color: var(--color-text-inverted);

  &:hover,
  &:focus {
    background-color: var(--color-accept-inverted);
    color: var(--color-text);
  }
`;

const SpanCancel = styled(Span)`
  background-color: var(--color-error);
  color: var(--color-text-inverted);

  &:hover,
  &:focus {
    background-color: var(--color-error-inverted);
    color: var(--color-text);
  }

  @media screen and (prefers-color-scheme: dark) {
    background-color: var(--color-link);
    color: var(--color-text-inverted);

    &:hover,
    &:focus {
      background-color: var(--color-link-hover);
      color: var(--color-text);
    }
  }
`;

const SpanDefaultThin = styled(SpanThin)`
  ${buttonDefaultThinStyle}
`;

export type ButtonProps = {
  action?: 'accept' | 'cancel' | 'default';
  children: React.ReactNode;
  onClick?: React.MouseEventHandler;
  thickness?: 'default' | 'thin';
};

function Button({
  action = 'default',
  children,
  onClick = undefined,
  thickness = 'default',
}: ButtonProps) {
  if (action === 'accept') {
    return <SpanAccept onClick={onClick}>{children}</SpanAccept>;
  }

  if (action === 'cancel') {
    return <SpanCancel onClick={onClick}>{children}</SpanCancel>;
  }

  if (thickness === 'thin') {
    return <SpanDefaultThin onClick={onClick}>{children}</SpanDefaultThin>;
  }

  return <SpanDefault onClick={onClick}>{children}</SpanDefault>;
}

export default Button;
