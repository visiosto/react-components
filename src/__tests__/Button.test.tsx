// Copyright (c) 2021 Visiosto oy
// Licensed under the MIT License

import React from 'react';

import Button from '../Button';

import renderWithProviders from '../util/test/renderWithProviders';

describe('Button component', () => {
  it('renders default colours correctly', () => {
    const { container, getByText } = renderWithProviders(<Button>Button</Button>);

    expect(container).toMatchSnapshot();

    expect(getByText('Button')).toBeInTheDocument();
  });

  it('renders accept colours correctly', () => {
    const { container, getByText } = renderWithProviders(<Button action="accept">Button</Button>);

    expect(container).toMatchSnapshot();

    expect(getByText('Button')).toBeInTheDocument();
  });

  it('renders cancel colours correctly', () => {
    const { container, getByText } = renderWithProviders(<Button action="cancel">Button</Button>);

    expect(container).toMatchSnapshot();

    expect(getByText('Button')).toBeInTheDocument();
  });

  it('renders default colours correctly on a thin button', () => {
    const { container, getByText } = renderWithProviders(<Button thickness="thin">Button</Button>);

    expect(container).toMatchSnapshot();

    expect(getByText('Button')).toBeInTheDocument();
  });

  it('renders accept colours correctly on a thin button', () => {
    const { container, getByText } = renderWithProviders(
      <Button action="accept" thickness="thin">
        Button
      </Button>,
    );

    expect(container).toMatchSnapshot();

    expect(getByText('Button')).toBeInTheDocument();
  });

  it('renders cancel colours correctly on a thin button', () => {
    const { container, getByText } = renderWithProviders(
      <Button action="cancel" thickness="thin">
        Button
      </Button>,
    );

    expect(container).toMatchSnapshot();

    expect(getByText('Button')).toBeInTheDocument();
  });
});
