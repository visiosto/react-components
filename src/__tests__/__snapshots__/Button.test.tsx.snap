// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component renders accept colours correctly 1`] = `
<div>
  <span
    class="sc-bdvvtL sc-hKwDye cQWVfy gRYswd"
  >
    Button
  </span>
</div>
`;

exports[`Button component renders accept colours correctly on a thin button 1`] = `
<div>
  <span
    class="sc-bdvvtL sc-hKwDye cQWVfy gRYswd"
  >
    Button
  </span>
</div>
`;

exports[`Button component renders cancel colours correctly 1`] = `
<div>
  <span
    class="sc-bdvvtL sc-eCImPb cQWVfy buYYcp"
  >
    Button
  </span>
</div>
`;

exports[`Button component renders cancel colours correctly on a thin button 1`] = `
<div>
  <span
    class="sc-bdvvtL sc-eCImPb cQWVfy buYYcp"
  >
    Button
  </span>
</div>
`;

exports[`Button component renders default colours correctly 1`] = `
<div>
  <span
    class="sc-bdvvtL sc-dkPtRN cQWVfy kLqXOg"
  >
    Button
  </span>
</div>
`;

exports[`Button component renders default colours correctly on a thin button 1`] = `
<div>
  <span
    class="sc-gsDKAQ sc-jRQBWg gxMheJ eVIhDg"
  >
    Button
  </span>
</div>
`;
